salsa-2022-tests/tests/debug.rs

```diff
@@ -44,15 +50,75 @@ fn input() {
     };
     let complex_struct = ComplexStruct::new(&db, input, not_salsa);
 
-    // default debug only includes identity fields
+    // debug includes all fields
     let actual = format!("{:?}", complex_struct.debug(&db));
-    let expected = expect!["ComplexStruct { [salsa id]: 0 }"];
-    expected.assert_eq(&actual);
-
-    // all fields
-    let actual = format!("{:?}", complex_struct.debug_all(&db));
     let expected = expect![[
         r#"ComplexStruct { [salsa id]: 0, my_input: MyInput { [salsa id]: 0, field: 22 }, not_salsa: NotSalsa { field: "it's salsa time" } }"#
     ]];
     expected.assert_eq(&actual);
 }
+
+#[salsa::tracked]
+fn leak_debug_string(db: &dyn Db, input: MyInput) -> String {
+    format!("{:?}", input.debug(db))
+}
+
+/// Test that field reads that occur as part of `Debug` are not tracked.
+/// Intentionally leaks the debug string.
+/// Don't try this at home, kids.
+#[test]
+fn untracked_dependencies() {
+    let mut db = Database::default();
+
+    let input = MyInput::new(&db, 22);
+
+    let s = leak_debug_string(&db, input);
+    expect![[r#"
+        "MyInput { [salsa id]: 0, field: 22 }"
+    "#]]
+    .assert_debug_eq(&s);
+
+    input.set_field(&mut db).to(23);
+
+    // check that we reuse the cached result for debug string
+    // even though the dependency changed.
+    let s = leak_debug_string(&db, input);
+    assert!(s.contains(", field: 22 }"));
+}
+
+#[salsa::tracked]
+#[customize(DebugWithDb)]
+struct DerivedCustom {
+    my_input: MyInput,
+    value: u32,
+}
+
+impl DebugWithDb<dyn Db + '_> for DerivedCustom {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>, db: &dyn Db) -> std::fmt::Result {
+        write!(
+            f,
+            "{:?} / {:?}",
+            self.my_input(db).debug(db),
+            self.value(db)
+        )
+    }
+}
```

Comment:
Nice: That's something I've come across too where some tracked structs have fields that are salsa ingredients and the "default" generated `DebugWithDb` implementation didn't emit them with `debug`. Thanks for making this change.